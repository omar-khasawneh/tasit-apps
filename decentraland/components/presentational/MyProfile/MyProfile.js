import React from "react";
import { StyleSheet, View } from "react-native";
import Colors from "@constants/Colors";
import PropTypes from "prop-types";
import { responsiveWidth } from "react-native-responsive-dimensions";
import MyProfileCreationStatusItem from "@presentational/MyProfileCreationStatusItem";
import ActionStatus from "@constants/ActionStatus";
import MyProfileProgress from "@presentational/MyProfileProgress";
import WalletButton from "@presentational/WalletButton";

// import { Linking } from "expo";
import AppLink from "react-native-app-link";

export default function MyProfile({ progress, creationSteps }) {
  const onConnect = async () => {
    try {
      let url = "tasit-wallet://transaction?hello=world&goodbye=now";

      console.info("Deep linking URL", url);

      // TODO: Change config to use test wallet app
      const config = {
        appName: "lyft",
        appStoreId: "529379082",
        appStoreLocale: "us",
        playStoreId: "me.lyft.android",
      };

      console.info("App store config", config);

      // TODO: Query for presence of the app using a separate function
      await AppLink.maybeOpenURL(url, config);
    } catch (error) {
      // handle error
      console.info("Error", error);
    }
  };

  return (
    <View style={styles.container}>
      <MyProfileProgress progress={progress} />
      <WalletButton onConnect={onConnect} />
      <View style={styles.actionItemsContainer}>
        {creationSteps.map(action => {
          const { name, status } = action;

          return (
            <MyProfileCreationStatusItem
              key={name}
              name={name}
              status={status}
            />
          );
        })}
      </View>
    </View>
  );
}

MyProfile.propTypes = {
  progress: PropTypes.number,
  creationSteps: PropTypes.arrayOf(
    PropTypes.shape({
      status: PropTypes.oneOf(Object.values(ActionStatus)),
      name: PropTypes.string,
    })
  ),
};

const styles = StyleSheet.create({
  actionItemsContainer: {
    flex: 5,
    flexDirection: "column",
    alignItems: "flex-start",
    justifyContent: "flex-start",
    paddingLeft: responsiveWidth(12),
  },
  container: {
    flex: 1,
    flexDirection: "column",
    backgroundColor: Colors.backgroundColor,
    alignItems: "flex-start",
    justifyContent: "flex-start",
  },
});
