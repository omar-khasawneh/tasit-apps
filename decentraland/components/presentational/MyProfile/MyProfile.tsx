import React from "react";
import { StyleSheet, View } from "react-native";
import Colors from "../../../constants/Colors";

import { responsiveWidth } from "react-native-responsive-dimensions";
import MyProfileCreationStatusItem from "../MyProfileCreationStatusItem";
import MyProfileProgress from "../MyProfileProgress";
import Button from "../Button";

import ActionStatus from "../../../types/ActionStatus";

const styles = StyleSheet.create({
  actionItemsContainer: {
    flex: 5,
    flexDirection: "column",
    alignItems: "flex-start",
    justifyContent: "flex-start",
    paddingLeft: responsiveWidth(12),
  },
  buttonContainer: {
    alignItems: "center",
    flex: 1,
    justifyContent: "center",
  },
  container: {
    flex: 1,
    flexDirection: "column",
    backgroundColor: Colors.backgroundColor,
    alignItems: "flex-start",
    justifyContent: "flex-start",
  },
});

interface CreationStepObject {
  status: ActionStatus;
  name: string;
}

interface MyProfileProps {
  progress: number;
  onClick: any; // TODO: Convert to function type
  creationSteps: CreationStepObject[];
}

const MyProfile: React.SFC<MyProfileProps> = ({
  progress,
  creationSteps,
  onClick,
}) => {
  return (
    <View style={styles.container}>
      <MyProfileProgress progress={progress} />
      <View style={styles.actionItemsContainer}>
        {creationSteps.map(action => {
          const { name, status } = action;

          return (
            <MyProfileCreationStatusItem
              key={name}
              name={name}
              status={status}
            />
          );
        })}
        {/* TODO: Break this component up into a smaller snapshot */}
        <View style={styles.buttonContainer}>
          <Button title="Connect wallet" onPress={onClick} />
        </View>
      </View>
    </View>
  );
};

export default MyProfile;
