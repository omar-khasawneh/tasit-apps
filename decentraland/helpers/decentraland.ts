// Note: Functions that will deal with fetch and prepare data from Decentraland contracts
import { addressesAreEqual } from "@helpers";

import AssetTypes from "@constants/AssetTypes";
const { ESTATE, PARCEL } = AssetTypes;

export const generateAssetFromId = async (
  estateContract,
  landContract,
  assetId,
  nftAddress
) => {
  const type = _getType(nftAddress, landContract, estateContract);
  let asset;

  if (type === ESTATE) {
    asset = await _generateEstateFromId(estateContract, assetId);
  } else if (type === PARCEL) {
    asset = await _generateParcelFromId(landContract, assetId);
  }

  return asset;
};

const _generateEstateFromId = async (estateContract, estateId) => {
  const id = `${estateId}`;
  const type = ESTATE;
  const name = await estateContract.getMetadata(id);
  const img = `https://api.decentraland.org/v1/estates/${id}/map.png`;

  const estate = { type, id, name, img };
  return estate;
};

const _generateParcelFromId = async (landContract, parcelId) => {
  const id = `${parcelId}`;
  const type = PARCEL;
  const namePromise = landContract.tokenMetadata(id);
  const coordsPromise = landContract.decodeTokenId(id);
  const [name, coords] = await Promise.all([namePromise, coordsPromise]);
  const [x, y] = coords;
  const img = `https://api.decentraland.org/v1/parcels/${x}/${y}/map.png`;

  const parcel = { type, id, name, img };
  return parcel;
};

const _getType = (nftAddress, landContract, estateContract) => {
  const isParcel = addressesAreEqual(nftAddress, landContract.getAddress());
  const isEstate = addressesAreEqual(nftAddress, estateContract.getAddress());

  if (!isParcel && !isEstate)
    throw new Error(
      `The sell order should have a parcel of land or an estate as the NFT.`
    );

  const type = isEstate ? ESTATE : PARCEL;
  return type;
};

export default {
  generateAssetFromId,
};
