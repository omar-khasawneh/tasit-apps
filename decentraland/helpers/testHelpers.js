import AssetTypes from "@constants/AssetTypes";
import { DONE, PENDING, MISSING } from "@constants/ActionStatus";
const { ESTATE, PARCEL } = AssetTypes;

export const parcel = {
  type: PARCEL,
  id: "0123456789",
  name: "Sample Parcel",
  img: "https://api.decentraland.org/v1/parcels/0/0/map.png",
};

export const estate = {
  type: ESTATE,
  id: "123",
  name: "Sample Estate",
  img: "https://api.decentraland.org/v1/estates/1/map.png",
};

export const estateWithoutName = {
  type: ESTATE,
  id: "321",
  name: "",
  img: "https://api.decentraland.org/v1/estates/1/map.png",
};

export const estateForSale = {
  id: "1",
  priceMana: "0",
  priceUSD: 0,
  asset: estate,
};

export const parcelForSale = {
  id: "2",
  priceMana: "0",
  priceUSD: 0,
  asset: parcel,
};

export const anAction = {
  getId: async () => "0x1234567890123456789012345678901234567890",
};

export const accountCreationSteps = [
  {
    name: "Account created",
    status: DONE,
  },
  {
    name: "ETH",
    status: PENDING,
  },
  {
    name: "MANA",
    status: MISSING,
  },
];

export default {
  estateForSale,
  parcelForSale,
  parcel,
  estate,
  anAction,
  estateWithoutName,
  accountCreationSteps,
};
