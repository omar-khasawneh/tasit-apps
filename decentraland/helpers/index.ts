import { Platform, Linking } from "react-native";
import { Toast } from "native-base";

import { Action, ContractBasedAccount, TasitContracts } from "tasit-sdk";

const { ConfigLoader } = Action;
import ProviderFactory from "tasit-action/dist/ProviderFactory";
import { createFromPrivateKey } from "tasit-account/dist/testHelpers/helpers";
import AccountCreationStatus from "../constants/AccountCreationStatus";

// TODO: Remove direct ethers dep from this app
import { ethers } from "ethers";

// Note: This is the 10th (#9 starting from index 0) account set up from
// the ganache-cli seed phrase the Tasit SDK is using
// As part of the migration, on all neteworks this account is set
// as the owner of the Gnosis Safe contract that is deployed
// Not a key controlling any funds or contracts on mainnet, of course:
const gnosisSafeOwnerPrivKey =
  "0x633a290bcdabb9075c5a4b3885c69ce64b4b0e6079eb929abb2ac9427c49733b";

const gnosisSafeOwner = createFromPrivateKey(gnosisSafeOwnerPrivKey);
const SMALL_AMOUNT = `${5e16}`; // 0.05
const HALF_MILLION = "500000000000000000000000";

const {
  NOT_STARTED,
  FUNDING_WITH_ETH,
  FUNDING_WITH_MANA_AND_APPROVING_MARKETPLACE,
  FUNDING_WITH_MANA,
  APPROVING_MARKETPLACE,
  READY_TO_USE,
} = AccountCreationStatus;

const ZERO = 0;

const loadConfig = (): void => {
  const tasitSdkConfig = require("../config/current");
  ConfigLoader.setConfig(tasitSdkConfig);
};

export const getNetworkName = (): string => {
  loadConfig();
  const provider = ProviderFactory.getProvider();
  const { _network: network } = provider;
  const networkName = !network ? "local" : network.name;
  return networkName;
};

interface GetAddressContract {
  getAddress: any; // TODO: Change me to a function type
}

interface MarketplaceContractObject extends GetAddressContract {
  setAccount: any; // TODO: Change me to a function type
  safeExecuteOrder: any; // TODO: Change me to a function type
}

interface EstateContractObject extends GetAddressContract {
  getFingerprint: any; // TODO: Change me to a function type
}

interface LandContractObject extends GetAddressContract {}

interface ManaContractObject extends GetAddressContract {
  setAccount: any; // TODO: Change me to a function type
  approve: any; // TODO: Change me to a function type
  balanceOf: any; // TODO: Change me to a function type
  allowance: any; // TODO: Change me to a function type
}

interface GnosisSafeContractObject {
  setAccount: any; // TODO: Change me to a function type
  setSigners: any; // TODO: Change me to a function type
  transferEther: any; // TODO: Change me to a function type
  transferERC20: any; // TODO: Change me to a function type
}

import { ActionObject } from "../types/ActionObject";

interface Contracts {
  marketplaceContract: MarketplaceContractObject;
  manaContract: ManaContractObject;
  gnosisSafeContract: GnosisSafeContractObject;
  estateContract: EstateContractObject;
  landContract: LandContractObject;
}

import { AccountObject } from "../types/AccountObject";

export const getContracts = (): Contracts => {
  const networkName = getNetworkName();

  // TODO: Once publishing an updated version of the Tasit SDK
  // is working again, remove this local logic for addresses.

  let theNetworkForAbiLookup = networkName;
  if (networkName === "rinkeby") {
    theNetworkForAbiLookup = "ropsten";
  }

  let {
    MANAToken,
    LANDProxy,
    EstateRegistry,
    Marketplace,
    GnosisSafe: GnosisSafeInfo,
  } = TasitContracts[theNetworkForAbiLookup];

  const rinkeby = {
    EstateRegistry: {
      address: "0x75acDF3f9C7E4f1141336401Da07265176530C19",
    },
    LANDProxy: {
      address: "0x74bFF5B8BaDd62DCf39E58216D23E03bA953cAe3",
    },
    LANDRegistry: {
      address: "0x8E91a85ce94430f83d81eAcCd3c018727e86bba0",
    },
    MANAToken: {
      address: "0x2A3e6dDa5b3451645f193194E377B18798420168",
    },
    Marketplace: {
      address: "0xE95Bc197Ba21B00BFbc8eCEBD4780be3427081be",
    },
    GnosisSafe: {
      address: "0x366dd6FdB8EA347A186e251e39bE1f12eF11cc8A",
    },
  };

  // TODO: Once publishing an updated version of the Tasit SDK
  // is working again, remove this local logic for addresses.
  if (networkName === "rinkeby") {
    // TODO: Add these addresses after running migration for rinkeby
    MANAToken.address = rinkeby.MANAToken.address;
    LANDProxy.address = rinkeby.LANDProxy.address;
    EstateRegistry.address = rinkeby.EstateRegistry.address;
    Marketplace.address = rinkeby.Marketplace.address;
    GnosisSafeInfo.address = rinkeby.GnosisSafe.address;
  }

  const { address: MANA_ADDRESS } = MANAToken;
  const { address: LAND_ADDRESS } = LANDProxy;
  const { address: ESTATE_ADDRESS } = EstateRegistry;
  const { address: MARKETPLACE_ADDRESS } = Marketplace;
  const { address: GNOSIS_SAFE_ADDRESS } = GnosisSafeInfo;

  const { ERC20, ERC721, Marketplace: MarketplaceContracts } = Action;
  const { Mana } = ERC20;
  const { Estate, Land } = ERC721;
  const { Decentraland: DecentralandMarketplace } = MarketplaceContracts;
  const { GnosisSafe } = ContractBasedAccount;

  const estateContract = new Estate(ESTATE_ADDRESS);
  const landContract = new Land(LAND_ADDRESS);
  const marketplaceContract = new DecentralandMarketplace(MARKETPLACE_ADDRESS);
  const manaContract = new Mana(MANA_ADDRESS);
  const gnosisSafeContract = new GnosisSafe(GNOSIS_SAFE_ADDRESS);

  const contracts = {
    estateContract,
    landContract,
    marketplaceContract,
    manaContract,
    gnosisSafeContract,
  };

  return contracts;
};

export const addressesAreEqual = (
  address1: { toUpperCase: () => void },
  address2: { toUpperCase: () => void }
): boolean => {
  return address1.toUpperCase() === address2.toUpperCase();
};

export const approveManaSpending = (
  fromAccount: AccountObject
): ActionObject => {
  const contracts = getContracts();
  const { manaContract, marketplaceContract } = contracts;
  const toAddress = marketplaceContract.getAddress();
  manaContract.setAccount(fromAccount);

  const action = manaContract.approve(toAddress, HALF_MILLION);
  return action;
};

// TODO: Consider making the amount to send a param of this function
export const fundAccountWithEthers = (accountAddress: string): ActionObject => {
  const { signingKey } = gnosisSafeOwner;
  const { address } = signingKey;
  console.log("Gnosis Safe master funding contract owner address", address);
  const contracts = getContracts();
  const { gnosisSafeContract } = contracts;
  gnosisSafeContract.setAccount(gnosisSafeOwner);
  gnosisSafeContract.setSigners([gnosisSafeOwner]);

  console.log("Amount to be transferred", SMALL_AMOUNT);

  const action = gnosisSafeContract.transferEther(accountAddress, SMALL_AMOUNT);
  console.info("Built the action to transfer ETH to", accountAddress);
  return action;
};

export const fundAccountWithEthersVanilla = async (
  accountAddress: string
): Promise<any> => {
  const { signingKey } = gnosisSafeOwner;
  const { privateKey } = signingKey;

  // We require a provider to send transactions
  const provider = ProviderFactory.getProvider();

  // let privateKey = "0x3141592653589793238462643383279502884197169399375105820974944592"
  let wallet = new ethers.Wallet(privateKey, provider);

  console.log("Amount to be transferred", SMALL_AMOUNT);

  console.info("Built the vanilla tx to transfer ETH to", accountAddress);

  let tx = {
    to: accountAddress,
    // We must pass in the amount as wei (1 ether = 1e18 wei), so we
    // use this convenience function to convert ether to wei.
    value: ethers.utils.parseEther("0.05"),
    // value: SMALL_AMOUNT,
  };

  // TODO: Too confusing: Change this
  console.info(
    "Sending from owner of funding Gnosis Safe rather than contract itself"
  );

  // let sendPromise = gnosisSafeOwner.sendTransaction(tx);
  let sendPromise = wallet.sendTransaction(tx);

  const txReturned = await sendPromise;
  console.log({ txReturned });
  return txReturned;
};

export const fundAccountWithMana = (accountAddress: string): ActionObject => {
  const contracts = getContracts();
  const { manaContract, gnosisSafeContract } = contracts;
  gnosisSafeContract.setAccount(gnosisSafeOwner);
  gnosisSafeContract.setSigners([gnosisSafeOwner]);

  const action = gnosisSafeContract.transferERC20(
    manaContract.getAddress(),
    accountAddress,
    HALF_MILLION
  );
  return action;
};

// Note: This could live inside of the Action class as `buildLink()` function
export const buildBlockchainUrlFromActionId = (actionId: any): string => {
  const networkName = getNetworkName();
  const transactionHash = actionId;
  const url = `https://${networkName}.etherscan.io/tx/${transactionHash}`;
  return url;
};

export const buildBlockchainUrlFromAddress = (address: string): string => {
  const networkName = getNetworkName();
  const url = `https://${networkName}.etherscan.io/address/${address}`;
  return url;
};

// More about Toast component: https://docs.nativebase.io/Components.html#toast-def-headref
const showToast = (msg: string): void =>
  Toast.show({ text: msg, duration: 3000, buttonText: "Okay" });

export const showFatalError = (msg: string): void => console.error(msg);
export const showError = (msg: string): void => showToast(`ERROR: ${msg}`);
export const showWarn = (msg: string): void => showToast(`WARN: ${msg}`);
export const showInfo = (msg: string): void => showToast(`${msg}`);

export const logInfo = (msg: string): void => console.info(msg);
export const logWarn = (msg: string): void => console.warn(msg);
export const logError = (msg: any): void => console.error(msg);

// Note: The root starter.ts script is not using this version
// of the checkBlockchain function, for ECMAScript version reasons
// The React Native app itself still is using this version, though
// TODO: Debug this further and remove the unneeded version / merge them

export const checkBlockchain = async (): Promise<boolean> => {
  logInfo("Checking blockchain");
  loadConfig();
  const provider = ProviderFactory.getProvider();
  try {
    await provider.getBlockNumber();
  } catch (err) {
    return false;
  }
  return true;
};

// Note: `toLocaleString` doesn't work on Android
// See more: https://github.com/facebook/react-native/issues/19410#issuecomment-434232762
export const formatNumber = (number: string | number): string => {
  if (typeof number === "string") {
    return number;
  }
  if (Platform.OS === "android") {
    // only android needs polyfill
    import("intl");
    import("intl/locale-data/jsonp/en-US");
  }

  // TODO: Handle internationalization for other regions
  const formatter = new Intl.NumberFormat("en-US");
  const formattedNumber = formatter.format(number);
  return formattedNumber;
};

interface Element {
  id: number | string;
}

export const toListIfNot = (itemOrList: Element | Element[]): Element[] =>
  Array.isArray(itemOrList) ? itemOrList : [itemOrList];

export const removeFromList = (
  list: { filter: (arg0: (e: any) => boolean) => object[] },
  toRemove: any
): object[] => {
  const elementsToRemove = toListIfNot(toRemove);
  const idsToRemove = elementsToRemove.map(e => e.id);
  return list.filter((e: { id: any }) => !idsToRemove.includes(e.id));
};

export const isInList = (
  list: { findIndex: (arg0: (e: any) => boolean) => number },
  toSearchFor: any
): boolean => {
  const elementsToFind = toListIfNot(toSearchFor);
  const idsToFind = elementsToFind.map(e => e.id);
  const result = list.findIndex((e: { id: any }) => !idsToFind.includes(e.id));
  if (!result) {
    return false;
  } else {
    return true;
  }
};

export const listsAreEqual = (first, second: object[]): boolean => {
  if (first.length !== second.length) return false;

  return (
    first.every((e: any) => second.includes(e)) &&
    second.every((e: any) => first.includes(e))
  );
};

// Update item from any list of objects having id as key field
export const updateListItem = (
  list: { map: (arg0: (item: any) => any) => object[] },
  toUpdateId: any,
  entriesToUpdate: any
): object[] => {
  return list.map((item: { id: any }) => {
    return item.id === toUpdateId ? { ...item, ...entriesToUpdate } : item;
  });
};

export const openURL = async (url: string): Promise<void> => {
  const supported = await Linking.canOpenURL(url);

  if (!supported) throw Error(`Can't handle url: ${url}`);

  try {
    await Linking.openURL(url);
  } catch (err) {
    throw Error(`Unable to open url: ${url}`);
  }
};

export const restoreCreationStateOfAccountFromBlockchain = async (account: {
  address: any;
}): Promise<object> => {
  const provider = ProviderFactory.getProvider();
  const { address } = account;

  const contracts: Contracts = getContracts();
  const { marketplaceContract, manaContract } = contracts;

  const ethersBalance = await provider.getBalance(address);
  const manaBalance = await manaContract.balanceOf(address);
  const allowance = await manaContract.allowance(
    address,
    marketplaceContract.getAddress()
  );

  let creationStatus = NOT_STARTED;
  const creationActions = {};

  const accountCreated = account !== null;
  const fundedWithEthers = `${ethersBalance}` !== `${ZERO}`;
  const fundedWithMana = `${manaBalance}` !== `${ZERO}`;
  const marketplaceWasApproved = `${allowance}` != `${ZERO}`;

  if (accountCreated) {
    creationStatus = FUNDING_WITH_ETH;
    creationActions[FUNDING_WITH_ETH] = null;
  }

  if (fundedWithEthers) {
    creationStatus = FUNDING_WITH_MANA_AND_APPROVING_MARKETPLACE;
    creationActions[APPROVING_MARKETPLACE] = null;
    creationActions[FUNDING_WITH_MANA] = null;
  }

  if (fundedWithMana && !marketplaceWasApproved) {
    creationStatus = APPROVING_MARKETPLACE;
  }

  if (!fundedWithMana && marketplaceWasApproved) {
    creationStatus = FUNDING_WITH_MANA;
  }

  if (fundedWithMana && marketplaceWasApproved) {
    creationStatus = READY_TO_USE;
  }

  return { creationStatus, creationActions };
};

export default {
  checkBlockchain,
  approveManaSpending,
  addressesAreEqual,
  showFatalError,
  showError,
  showWarn,
  showInfo,
  logInfo,
  logWarn,
  logError,
  fundAccountWithEthers,
  fundAccountWithEthersVanilla,
  fundAccountWithMana,
  getContracts,
  formatNumber,
  removeFromList,
  isInList,
  listsAreEqual,
  openURL,
  getNetworkName,
  buildBlockchainUrlFromActionId,
  buildBlockchainUrlFromAddress,
  restoreCreationStateOfAccountFromBlockchain,
  updateListItem,
  toListIfNot,
};
