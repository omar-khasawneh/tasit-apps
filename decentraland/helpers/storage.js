import { SecureStore } from "expo";
import { AsyncStorage } from "react-native";
import { createFromPrivateKey } from "tasit-account/dist/testHelpers/helpers";

// Storage keys
const EPHEMERAL_ACCOUNT_PRIV_KEY = "EPHEMERAL_ACCOUNT_PRIV_KEY";
const MY_ASSETS_LIST = "MY_ASSETS_LIST";
const EPHEMERAL_ACCOUNT_CREATION_STATUS = "EPHEMERAL_ACCOUNT_CREATION_STATUS";
const EPHEMERAL_ACCOUNT_CREATION_ACTIONS = "EPHEMERAL_ACCOUNT_CREATION_ACTIONS";
const APP_IS_FIRST_USE = "APP_IS_FIRST_USE";

export const storeIsFirstUse = async isFirstUse => {
  const strIsFirstUse = _toString(isFirstUse);
  await _storeData(APP_IS_FIRST_USE, strIsFirstUse, false);
};

export const retrieveIsFirstUse = async () => {
  const strIsFirstUse = await _retrieveData(APP_IS_FIRST_USE);
  const isFirstUse = _fromString(strIsFirstUse);
  if (isFirstUse === null) return true;
  return isFirstUse;
};

export const storeAccountCreationActions = async creationActions => {
  const strCreationActions = _toString(creationActions);
  await _storeData(
    EPHEMERAL_ACCOUNT_CREATION_ACTIONS,
    strCreationActions,
    false
  );
};

export const retrieveAccountCreationActions = async () => {
  const strCreationActions = await _retrieveData(
    EPHEMERAL_ACCOUNT_CREATION_ACTIONS
  );
  let creationActions = _fromString(strCreationActions);
  return creationActions;
};

export const storeAccountCreationStatus = async status => {
  await _storeData(EPHEMERAL_ACCOUNT_CREATION_STATUS, status, false);
};

export const retrieveAccountCreationStatus = async () => {
  const status = await _retrieveData(EPHEMERAL_ACCOUNT_CREATION_STATUS);
  return status;
};

export const storeAccount = async account => {
  const { privateKey } = account;
  await _storeData(EPHEMERAL_ACCOUNT_PRIV_KEY, privateKey, true);
};

export const retrieveAccount = async () => {
  let account = null;

  const privateKey = await _retrieveData(EPHEMERAL_ACCOUNT_PRIV_KEY);
  if (privateKey != null) account = createFromPrivateKey(privateKey);

  return account;
};

export const storeMyAssets = async myAssets => {
  const strMyAssets = _toString(myAssets);
  await _storeData(MY_ASSETS_LIST, strMyAssets, false);
};

export const retrieveMyAssets = async () => {
  const strMyAssets = await _retrieveData(MY_ASSETS_LIST);
  let myAssets = _fromString(strMyAssets);
  if (!myAssets) myAssets = [];
  return myAssets;
};

const _toString = obj => {
  try {
    return JSON.stringify(obj);
  } catch {
    throw Error(`Unable to parse object to a JSON string.`);
  }
};

const _fromString = string => {
  try {
    return JSON.parse(string);
  } catch {
    throw Error(`Unable to parse JSON string to an object.`);
  }
};

// Note: the value should be a string
const _storeData = async (key, value, securely) => {
  try {
    if (securely) {
      // More about options:
      // https://docs.expo.io/versions/latest/sdk/securestore/#securestoresetitemasynckey-value-options
      const options = { keychainAccessible: SecureStore.WHEN_UNLOCKED };
      await SecureStore.setItemAsync(key, value, options);
    } else {
      await AsyncStorage.setItem(key, value);
    }
  } catch (error) {
    throw Error(`Unable to ${securely ? "securely" : ""} store data.`);
  }
};

const _retrieveData = async key => {
  try {
    let value = await AsyncStorage.getItem(key);
    if (value === null) value = await SecureStore.getItemAsync(key);

    return value;
  } catch (error) {
    throw Error(`Unable to retrieve data from storage.`);
  }
};

export default {
  storeAccount,
  retrieveAccount,
  storeMyAssets,
  retrieveMyAssets,
  storeAccountCreationStatus,
  retrieveAccountCreationStatus,
  storeAccountCreationActions,
  retrieveAccountCreationActions,
  storeIsFirstUse,
  retrieveIsFirstUse,
};
