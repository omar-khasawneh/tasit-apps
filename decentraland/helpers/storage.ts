// Note: If a piece of data wasn't found, returning null instead of default state of the
// property to avoid coupling with redux reducers
import * as SecureStore from "expo-secure-store";
import { AsyncStorage } from "react-native";
import { createFromPrivateKey } from "tasit-account/dist/testHelpers/helpers";
import { logWarn } from "@helpers";

// Storage keys
const EPHEMERAL_ACCOUNT_PRIV_KEY = "EPHEMERAL_ACCOUNT_PRIV_KEY";
const MY_ASSETS_LIST = "MY_ASSETS_LIST";
const EPHEMERAL_ACCOUNT_CREATION_STATUS = "EPHEMERAL_ACCOUNT_CREATION_STATUS";
const EPHEMERAL_ACCOUNT_CREATION_ACTIONS = "EPHEMERAL_ACCOUNT_CREATION_ACTIONS";
const IS_FIRST_APP_USE = "IS_FIRST_APP_USE";
const USER_ACTIONS = "USER_ACTIONS";

const _clearData = async (key): Promise<any> => {
  try {
    await AsyncStorage.removeItem(key);
  } catch (error) {
    logWarn(`Unable to delete data (key = ${key}) from storage.`);
  }
  try {
    await SecureStore.deleteItemAsync(key);
  } catch (error) {
    logWarn(`Unable to delete data key = ${key} from secure storage.`);
  }
};

export const clearAllStorage = async () => {
  await _clearData(IS_FIRST_APP_USE);
  await _clearData(EPHEMERAL_ACCOUNT_PRIV_KEY);
  await _clearData(USER_ACTIONS);
  await _clearData(EPHEMERAL_ACCOUNT_CREATION_ACTIONS);
  await _clearData(EPHEMERAL_ACCOUNT_CREATION_STATUS);
  await _clearData(MY_ASSETS_LIST);
};

const _toString = (obj): string => {
  if (obj === null) return null;

  try {
    return JSON.stringify(obj);
  } catch {
    throw Error(`Unable to parse object to a JSON string.`);
  }
};

const _fromString = (string): any => {
  try {
    return JSON.parse(string);
  } catch {
    throw Error(`Unable to parse JSON string to an object.`);
  }
};

// Note: the value should be a string
const _storeData = async (key, value, securely): Promise<any> => {
  try {
    if (securely) {
      // More about options:
      // https://docs.expo.io/versions/latest/sdk/securestore/#securestoresetitemasynckey-value-options
      const options = { keychainAccessible: SecureStore.WHEN_UNLOCKED };
      await SecureStore.setItemAsync(key, value, options);
    } else {
      await AsyncStorage.setItem(key, value);
    }
  } catch (error) {
    throw Error(`Unable to ${securely ? "securely" : ""} store data.`);
  }
};

export const storeUserActions = async (userActions): Promise<any> => {
  const strUserActions = _toString(userActions);
  await _storeData(USER_ACTIONS, strUserActions, false);
};

const _retrieveData = async (key): Promise<any> => {
  try {
    let value = await AsyncStorage.getItem(key);
    if (value === null) value = await SecureStore.getItemAsync(key);

    return value;
  } catch (error) {
    throw Error(`Unable to retrieve data from storage.`);
  }
};

export const retrieveUserActions = async (): Promise<object> => {
  const strUserActions = await _retrieveData(USER_ACTIONS);
  const userActions = _fromString(strUserActions);
  return userActions;
};

export const storeIsFirstUse = async (isFirstUse): Promise<any> => {
  const strIsFirstUse = _toString(isFirstUse);
  await _storeData(IS_FIRST_APP_USE, strIsFirstUse, false);
};

export const retrieveIsFirstUse = async (): Promise<boolean> => {
  const strIsFirstUse = await _retrieveData(IS_FIRST_APP_USE);
  const isFirstUse = _fromString(strIsFirstUse);
  if (isFirstUse === null) return true;
  return isFirstUse;
};

export const storeAccountCreationActions = async (
  creationActions
): Promise<any> => {
  const strCreationActions = _toString(creationActions);
  await _storeData(
    EPHEMERAL_ACCOUNT_CREATION_ACTIONS,
    strCreationActions,
    false
  );
};

export const retrieveAccountCreationActions = async (): Promise<object[]> => {
  const strCreationActions = await _retrieveData(
    EPHEMERAL_ACCOUNT_CREATION_ACTIONS
  );
  const creationActions = _fromString(strCreationActions);
  return creationActions;
};

export const storeAccountCreationStatus = async (status): Promise<any> => {
  await _storeData(EPHEMERAL_ACCOUNT_CREATION_STATUS, status, false);
};

export const retrieveAccountCreationStatus = async (): Promise<string> => {
  const status = await _retrieveData(EPHEMERAL_ACCOUNT_CREATION_STATUS);
  return status;
};

export const storeAccount = async (account): Promise<any> => {
  const privateKey = account !== null ? account.privateKey : null;
  await _storeData(EPHEMERAL_ACCOUNT_PRIV_KEY, privateKey, true);
};

export const retrieveAccount = async (): Promise<object> => {
  let account = null;

  const privateKey = await _retrieveData(EPHEMERAL_ACCOUNT_PRIV_KEY);
  if (privateKey != null) account = createFromPrivateKey(privateKey);

  return account;
};

export const storeMyAssets = async (myAssets): Promise<any> => {
  const strMyAssets = _toString(myAssets);
  await _storeData(MY_ASSETS_LIST, strMyAssets, false);
};

export const retrieveMyAssets = async (): Promise<object[]> => {
  const strMyAssets = await _retrieveData(MY_ASSETS_LIST);
  const myAssets = _fromString(strMyAssets);
  return myAssets;
};

export default {
  storeAccount,
  retrieveAccount,
  storeMyAssets,
  retrieveMyAssets,
  storeAccountCreationStatus,
  retrieveAccountCreationStatus,
  storeAccountCreationActions,
  retrieveAccountCreationActions,
  storeIsFirstUse,
  retrieveIsFirstUse,
  retrieveUserActions,
  storeUserActions,
  clearAllStorage,
};
