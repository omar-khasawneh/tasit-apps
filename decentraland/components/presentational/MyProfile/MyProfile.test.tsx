import React from "react";
import { shallow } from "enzyme";
import MyProfile from "../MyProfile";
import { accountCreationSteps } from "../../../helpers/testHelpers";

describe("MyProfile", () => {
  it("renders the component", () => {
    const progress = 0.5;
    const onClick = (): void => {};
    expect(
      shallow(
        <MyProfile
          onClick={onClick}
          progress={progress}
          creationSteps={accountCreationSteps}
        />
      )
    ).toMatchSnapshot();
  });
});
