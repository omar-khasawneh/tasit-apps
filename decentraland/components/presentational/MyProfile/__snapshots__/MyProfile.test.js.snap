// Jest Snapshot v1, https://goo.gl/fbAQLP

/* eslint-disable jest/no-large-snapshots */
exports[`MyProfile renders the component 1`] = `
<Component
  style={
    Object {
      "alignItems": "flex-start",
      "backgroundColor": "black",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
    }
  }
>
  <MyProfileProgress
    progress={0.5}
  />
  <Component
    style={
      Object {
        "alignItems": "flex-start",
        "flex": 5,
        "flexDirection": "column",
        "justifyContent": "flex-start",
        "paddingLeft": 90,
      }
    }
  >
    <MyProfileCreationStatusItem
      key="Account created"
      name="Account created"
      status="done"
    />
    <MyProfileCreationStatusItem
      key="ETH"
      name="ETH"
      status="pending"
    />
    <MyProfileCreationStatusItem
      key="MANA"
      name="MANA"
      status="missing"
    />
  </Component>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Button
      onPress={[Function]}
      title="Connect wallet"
    />
  </Component>
</Component>
`;
